src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Tds/Tokens/LoginAck/LoginAckTokenParser.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClientX.IO;
+
+namespace Microsoft.Data.SqlClientX.Tds.Tokens.LoginAck
+{
+    internal class LoginAckTokenParser : TokenParser
+    {
+        public override async ValueTask<Token> ParseAsync(TokenType tokenType, TdsStream tdsStream, bool isAsync, CancellationToken ct)
+        {
+            _ = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct).ConfigureAwait(false); // length
+
+            byte type = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+            SqlInterfaceType interfaceType = LoginAckTokenParser.GetSqlInterfaceType(type);
+
+            uint version = await tdsStream.TdsReader.ReadUInt32BEAsync(isAsync, ct).ConfigureAwait(false);
+            TdsVersion tdsVersion = LoginAckTokenParser.GetTdsVersion(version);
+            tdsStream.TdsVersion = tdsVersion;
+
+            string progName = await tdsStream.TdsReader.ReadBVarCharAsync(isAsync, ct).ConfigureAwait(false);
+            byte major = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
```

Comment:
I've read the disclaimer in the PR regarding benchmarks, I still have a few questions (this is generic, I'll just take this token as an example):
1. How many of the tokens (any token) exist per typical sql operation, I'm a bit concerned about the amount of objects created
2. Are there any tokens with only a single Read operation? because the whole ValueTask as return type is kinda moot, if there is always an await in there, the same applies to the ValueTasks in TdsReader.

Honestly, the amount of awaits here is frightening, we're talking about a protocol parser after all, there should be 0 (maybe 1) awaits in here besides a possible await for  VarCharAsync read as you know the minimum bytes required to build this token.

In a different PR I was told that the C# pipeline concept does not really apply to the usage of TDS, but with the token parser concept, can that previous explanation be revisited?

Reply:
Delegating this discussion to https://github.com/dotnet/SqlClient/discussions/2689, I'd request posting further comments on this topic there. We are considerate of the amount of awaits, and this is something we need to incrementally rework in the I/O design, using ByteBuffer or a similar solution by working with Buffer directly.

Reply:
I was going to wait for the PR to merge before I started to look at benchmarks, but to keep performance comments together, [this](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/490e563d-cc6e-4c86-bb95-ef0186b98032) is the relevant section of the TDS specification for a LOGINACK token.

My first instinct would be to look at the token's Length field, then to fill a buffer with a sensible number of bytes up to that length, and allow the token to interpret it. I'm not sure whether there's much value in this though: the TDS reader already stores a 4KB TDS packet in memory, and it'll read synchronously from this packet if it can. An extra buffer on top of this seems like overkill. There might be value in providing a way to instruct the TDS reader to guarantee that a certain number of bytes are available in its buffer though, as this'd render the token parsing synchronous.

I agree with your point about creating one object per token; it needs benchmarking. We're not yet at the point of executing SQL statements, but I'm conscious that each row in a result set is also a ROW token...

Reply:
We're on the same page!
I agree with not creating any extra buffer and reading off the stream directly. The solution we need is certainly being able to ensure buffer can be prefilled, which is what we need to achieve somewhere outside the token stream handling in I/O layer. Because it doesn't exist as of now, these are async calls, I would imagine us to convert the 'ParseAsync' to a sync API in future, ensuring buffer is prefilled, and reading off it synchronously.

So please don't take these PRs as final designs as of now, these are incremental changes :)